pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package traps
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/gosnmp/gosnmp"
+)
+
+// Formatter is an interface to extract and format raw SNMP Traps
+type Formatter interface {
+	FormatPacket(packet *SnmpPacket) ([]byte, error)
+	GetTags(packet *SnmpPacket) []string
+}
+
+// JSONFormatter is a Formatter implementation that transforms Traps into JSON
+type JSONFormatter struct {
+	oidResolver OIDResolver
+	namespace   string
+}
+
+type trapVariable struct {
+	OID     string      `json:"oid"`
+	VarType string      `json:"type"`
+	Value   interface{} `json:"value"`
+}
+
+const (
+	sysUpTimeInstanceOID = "1.3.6.1.2.1.1.3.0"
+	snmpTrapOID          = "1.3.6.1.6.3.1.1.4.1.0"
+)
+
+// NewJSONFormatter creates a new JSONFormatter instance with an optional OIDResolver variable.
+func NewJSONFormatter(oidResolver OIDResolver) JSONFormatter {
+	namespace := GetNamespace()
+	if oidResolver == nil {
+		return JSONFormatter{NoOpOIDResolver{}, namespace}
+	}
+	return JSONFormatter{oidResolver, namespace}
+}
+
+// FormatPacket converts a raw SNMP trap packet to a FormattedSnmpPacket containing the JSON data and the tags to attach
+func (f JSONFormatter) FormatPacket(packet *SnmpPacket) ([]byte, error) {
+	var formattedData map[string]interface{}
+	var err error
+	if packet.Content.Version == gosnmp.Version1 {
+		formattedData = f.formatV1Trap(packet.Content)
+	} else {
+		formattedData, err = f.formatTrap(packet.Content)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return json.Marshal(formattedData)
+}
+
+// GetTags returns a list of tags associated to an SNMP trap packet.
+func (f JSONFormatter) GetTags(packet *SnmpPacket) []string {
+	return []string{
+		fmt.Sprintf("snmp_version:%s", formatVersion(packet.Content)),
```

Comment:
Similar to https://github.com/DataDog/datadog-agent/pull/10934#discussion_r819702696

Do we need to report version ?

Reply:
Same answer, didn't want to change it in this PR. Do you think we should remove it?
The version doesn't provide I agree but it doesn't harm to have it